File: 프로그래머스/2/150369. 택배 배달과 수거하기/택배 배달과 수거하기.py
def solution(cap, n, deliveries, pickups):
    answer = 0
    while True:
        if not deliveries and not pickups:
            break
        for i in range(len(deliveries) - 1, -1, -1):
            if deliveries[i] == 0:
                deliveries.pop(i)
            else:
                break
        for i in range(len(pickups) - 1, -1, -1):
            if pickups[i] == 0:
                pickups.pop(i)
            else:
                break
        answer += 2* (max(len(deliveries), len(pickups)))
        cnt = cap
        
        for i in range(len(deliveries) - 1, -1, -1):
            if deliveries[i] <= cnt:
                cnt -= deliveries[i]
                deliveries.pop(i)
            else:
                deliveries[i] -= cnt
                break
        cnt = cap
        for i in range(len(pickups) - 1, -1, -1):
            if pickups[i] <= cnt:
                cnt -= pickups[i]
                pickups.pop(i)
            else:
                pickups[i] -= cnt
                break
    return answer
                
                

File: 프로그래머스/2/150369. 택배 배달과 수거하기/택배 배달과 수거하기.java
import java.util.*;

class Solution {
    public long solution(int cap, int n, int[] deliveries, int[] pickups) {
        long answer = 0;
        
        PriorityQueue<int[]> deliveryPQ = new PriorityQueue<>((a, b) -> Integer.compare(b[0], a[0]));
        PriorityQueue<int[]> pickupPQ = new PriorityQueue<>((a, b) -> Integer.compare(b[0], a[0]));
        
        for (int i = 0; i < n; i++) {
            int cnt = deliveries[i];
            if (cnt == 0) continue;
            deliveryPQ.add(new int[] {i + 1, cnt});
        }
        
        for (int i = 0; i < n; i++) {
            int cnt = pickups[i];
            if (cnt == 0) continue;
            pickupPQ.add(new int[] {i + 1, cnt});
        }
        
        while (true) {
            if (deliveryPQ.isEmpty() && pickupPQ.isEmpty()) break;
            
            int cur_cap = cap;
            int delivery_dist = 0;
            while(!deliveryPQ.isEmpty()) {
                int[] cur_arr = deliveryPQ.poll();
                
                delivery_dist = Math.max(delivery_dist, cur_arr[0]);
                if (cur_arr[1] > cur_cap) {
                    cur_arr[1] -= cur_cap;
                    cur_cap = 0;
                    deliveryPQ.add(cur_arr);
                } else {
                    cur_cap -= cur_arr[1];
                }
                
                if (cur_cap == 0) break;
            }
            
            cur_cap = cap;
            int pick_dist = delivery_dist;
            while(!pickupPQ.isEmpty()) {
                int[] cur_arr = pickupPQ.poll();
                      
                pick_dist = Math.max(pick_dist, cur_arr[0]);
                if (cur_arr[1] > cur_cap) {
                    cur_arr[1] -= cur_cap;
                    cur_cap = 0;
                    pickupPQ.add(cur_arr);
                } else {
                    cur_cap -= cur_arr[1];
                }
                if (cur_cap == 0) break;
            }
            
            answer += Math.max(delivery_dist, pick_dist) * 2;


        }
        
        return answer;
    }
}